src/smart-escrow/SmartEscrow.sol

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import "@openzeppelin/contracts/access/Ownable2Step.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/// @dev The error is thrown when an address is not set.
+error AddressIsZeroAddress();
+
+/// @dev The error is thrown when the start timestamp is zero.
+error StartTimeIsZero();
+
+/// @dev The error is thrown when the start timestamp is greater than the end timestamp.
+error StartTimeAfterEndTime(uint256 startTimestamp, uint256 endTimestamp);
+
+/// @dev The error is thrown when the vesting period is zero.
+error VestingPeriodIsZeroSeconds();
+
+/// @dev The error is thrown when the caller of the method is not the expected owner.
+error CallerIsNotOwner(address caller, address owner);
+
+/// @dev The error is thrown when the contract is terminated, when it should not be.
+error ContractIsTerminated();
+
+/// @dev The error is thrown when the contract is not terminated, when it should be.
+error ContractIsNotTerminated();
+
+/**
+ * @title SmartEscrow contract
+ * @dev Contract to handle payment of OP tokens over a period of vesting with
+ * the ability to terminate the contract.
+ * This contract is inspired by OpenZeppelin's VestingWallet contract, but had sufficiently
+ * different requirements to where inheriting did not make sense.
+ */
+contract SmartEscrow is Ownable2Step {
+    event OPTransfered(uint256 amount, address indexed recipient);
+    event BeneficiaryOwnerUpdated(address indexed oldOwner, address indexed newOwner);
+    event BeneficiaryUpdated(address indexed oldBeneficiary, address indexed newBeneficiary);
+    event ContractTerminated();
+
+    IERC20 public constant OP_TOKEN = IERC20(0x4200000000000000000000000000000000000042);
+
+    address public beneficiaryOwner;
+    address public beneficiary;
+    uint256 public released;
+    uint256 public immutable start;
+    uint256 public immutable end;
+    uint256 public immutable vestingPeriod;
+    uint256 public immutable initialTokens;
+    uint256 public immutable  vestingEventTokens;
+    bool public contractTerminated;
+
+    /**
+     * @dev Set initial parameters.
+     * @param beneficiaryOwnerAddress Address which can update the beneficiary address.
+     * @param beneficiaryAddress Address which receives tokens that have vested.
+     * @param escrowOwner Address which can terminate the contract.
+     * @param startTimestamp Timestamp of the start of vesting period (or the cliff, if there is one).
+     * @param endTimestamp Timestamp of the end of the vesting period.
+     * @param vestingPeriodSeconds Period of time between each vesting event in seconds.
+     * @param numInitialTokens Number of OP tokens which vest at start time.
+     * @param numVestingEventTokens Number of OP tokens which vest upon each vesting event.
+     */
```

Comment:
Would be nice to have all comment styling consistent - preferably solidity natspec.